File: data/repos/genai_toolbox/content/docs/en/samples/pre_post_processing/python/adk/agent.py
import asyncio
from datetime import datetime
from typing import Any, Dict, Optional
from copy import deepcopy

from google.adk import Agent
from google.adk.apps import App
from google.adk.runners import Runner
from google.adk.sessions.in_memory_session_service import InMemorySessionService
from google.adk.tools.tool_context import ToolContext
from google.genai import types
from toolbox_adk import CredentialStrategy, ToolboxToolset, ToolboxTool

SYSTEM_PROMPT = """
  You're a helpful hotel assistant. You handle hotel searching, booking and
  cancellations. When the user searches for a hotel, mention it's name, id,
  location and price tier. Always mention hotel ids while performing any
  searches. This is very important for any operations. For any bookings or
  cancellations, please provide the appropriate confirmation. Be sure to
  update checkin or checkout dates if mentioned by the user.
  Don't ask for confirmations from the user.
"""


# Pre processing
async def enfore_business_rules(
    tool: ToolboxTool, args: Dict[str, Any], tool_context: ToolContext
) -> Optional[Dict[str, Any]]:
    """
    Callback fired before a tool is executed.
    Enforces business logic: Max stay duration is 14 days.
    """
    tool_name = tool.name
    print(f"POLICY CHECK: Intercepting '{tool_name}'")

    if tool_name == "update-hotel" and "checkin_date" in args and "checkout_date" in args:
        start = datetime.fromisoformat(args["checkin_date"])
        end = datetime.fromisoformat(args["checkout_date"])
        duration = (end - start).days

        if duration > 14:
            print("BLOCKED: Stay too long")
            return {"result": "Error: Maximum stay duration is 14 days."}
    return None


# Post processing
async def enrich_response(
    tool: ToolboxTool,
    args: Dict[str, Any],
    tool_context: ToolContext,
    tool_response: Any,
) -> Optional[Any]:
    """
    Callback fired after a tool execution.
    Enriches response for successful bookings.
    """
    if isinstance(tool_response, dict):
        result = tool_response.get("result", "")
    elif isinstance(tool_response, str):
        result = tool_response
    else:
        return None

    tool_name = tool.name
    if isinstance(result, str) and "Error" not in result:
        if tool_name == "book-hotel":
            loyalty_bonus = 500
            enriched_result = f"Booking Confirmed!\n You earned {loyalty_bonus} Loyalty Points with this stay.\n\nSystem Details: {result}"

            if isinstance(tool_response, dict):
                modified_response = deepcopy(tool_response)
                modified_response["result"] = enriched_result
                return modified_response
            else:
                return enriched_result
    return None


async def run_chat_turn(
    runner: Runner, session_id: str, user_id: str, message_text: str
):
    """Executes a single chat turn and prints the interaction."""
    print(f"\nUSER: '{message_text}'")
    response_text = ""
    async for event in runner.run_async(
        user_id=user_id,
        session_id=session_id,
        new_message=types.Content(role="user", parts=[types.Part(text=message_text)]),
    ):
        if event.content and event.content.parts:
            for part in event.content.parts:
                if part.text:
                    response_text += part.text

    print(f"AI: {response_text}")


async def main():
    toolset = ToolboxToolset(
        server_url="http://127.0.0.1:5000",
        toolset_name="my-toolset",
        credentials=CredentialStrategy.toolbox_identity(),
    )
    tools = await toolset.get_tools()
    root_agent = Agent(
        name="root_agent",
        model="gemini-2.5-flash",
        instruction=SYSTEM_PROMPT,
        tools=tools,
        # add any pre and post processing callbacks
        before_tool_callback=enfore_business_rules,
        after_tool_callback=enrich_response,
    )
    app = App(root_agent=root_agent, name="my_agent")
    runner = Runner(app=app, session_service=InMemorySessionService())
    session_id = "test-session"
    user_id = "test-user"
    await runner.session_service.create_session(
        app_name=app.name, user_id=user_id, session_id=session_id
    )

    # First turn: Successful booking
    await run_chat_turn(runner, session_id, user_id, "Book hotel with id 3.")
    print("-" * 50)
    # Second turn: Policy violation (stay > 14 days)
    await run_chat_turn(
        runner,
        session_id,
        user_id,
        "Book a hotel with id 5 with checkin date 2025-01-18 and checkout date 2025-02-10",
    )
    await toolset.close()


if __name__ == "__main__":
    asyncio.run(main())

File: public/data/repos/genai_toolbox/content/docs/en/samples/pre_post_processing/js/langchain/agent.js
import { ToolboxClient } from "@toolbox-sdk/core";
import { ChatGoogleGenerativeAI } from "@langchain/google-genai";
import { createAgent, createMiddleware, ToolMessage } from "langchain";
import { tool } from "@langchain/core/tools";
import { fileURLToPath } from "url";
import process from "process";

const systemPrompt = `
You're a helpful hotel assistant. You handle hotel searching, booking and
cancellations. When the user searches for a hotel, mention it's name, id,
location and price tier. Always mention hotel ids while performing any
searches. This is very important for any operations. For any bookings or
cancellations, please provide the appropriate confirmation. Be sure to
update checkin or checkout dates if mentioned by the user.
Don't ask for confirmations from the user.
`;

const GOOGLE_API_KEY = process.env.GOOGLE_API_KEY || 'your-api-key'; // Replace it with your API key

const businessRulesMiddleware = createMiddleware({
  name: "BusinessRules",
  wrapToolCall: async (request, handler) => {
    const toolName = request.toolCall.name;
    const toolArgs = request.toolCall.args;
    console.log(`POLICY CHECK: Intercepting '${toolName}' running with args ${JSON.stringify(toolArgs)}`);
    if (toolName === "update-hotel" && toolArgs.checkin_date && toolArgs.checkout_date) {
      try {
        const start = new Date(toolArgs.checkin_date);
        const end = new Date(toolArgs.checkout_date);
        const duration = (end - start) / (1000 * 60 * 60 * 24); // days

        if (duration > 14) {
          console.log("BLOCKED: Stay too long");
          return ToolMessage({content:'Error: Maximum stay duration is 14 days.', status:"error"})
        }
      } catch (e) {
        // Ignore invalid dates
      }
    }
    return handler(request);
  }
});

const enrichmentMiddleware = createMiddleware({
  name: "Enrichment",
  wrapToolCall: async (request, handler) => {
    const result = await handler(request);
    const toolName = request.toolCall.name;
    
    let content = result;
    if (typeof result === 'object' && result !== null && result.content) {
        content = result.content;
    }
    if (toolName === "book-hotel" && typeof content === 'string' && !content.includes("Error")) {
        const loyaltyBonus = 500;
        const enrichedContent = `Booking Confirmed!\n You earned ${loyaltyBonus} Loyalty Points with this stay.\n\nSystem Details: ${content}`;
        if (typeof result === 'object' && result !== null) {
            result.content = enrichedContent;
            return result;
        }
        return enrichedContent;
    }
    return result;
  }
});

const queries = [
  "Book hotel with id 3.",
  "Update my hotel with id 3 with checkin date 2025-01-18 and checkout date 2025-02-10"
];

async function main() {
  const client = new ToolboxClient("http://127.0.0.1:5000");
  const rawTools = await client.loadToolset("my-toolset");
  const tools = rawTools
    .map(t => tool(t, {
        name: t.getName(),
        description: t.getDescription(),
        schema: t.getParamSchema()
  }));

  const model = new ChatGoogleGenerativeAI({
    model: "gemini-2.5-flash",
  });

  const agent = createAgent({ 
    model: model, 
    tools: tools, 
    systemPrompt: systemPrompt, 
    middleware: [businessRulesMiddleware, enrichmentMiddleware] 
  });

  for (const query of queries) {
    console.log(`\nUSER: '${query}'`);
    const result = await agent.invoke({
      messages: [
        { role: "user", content: query},
      ],
    });
    console.log("-".repeat(50));
    console.log(`AI: ${result.messages[result.messages.length-1].content}`);
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main();
}

export { main };
